### ISSUE TITLE
Proposal: Have the ability to override the wp-config.php path

### ISSUE THREAD

**Issue by jonathanbardo** [wp-cli/wp-cli#5120]
I would like to propose to add the ability to provide our own config override path based on an environment variable (WP_CONFIG_PATH). This would solve a lot of headaches for managed hosts that rely on wp-cli but can't trust the wp-config that the user can play with. (For example, when the WordPress site loads fine on the server but somehow the wp-config.php custom php code prevents wp-cli from loading correctly).

**Comment by fjarrett** [comment]
I've been discussing this a lot with @jonathanbardo internally here at GoDaddy as we work to solve some common problems we encounter. The chief among them is customers breaking their own config files with bad syntax (or getting hacked) causing our WP-CLI calls to fail.

The solution proposed here aims to solve for this use case (and potentially others) where admins already possess DB credentials and need to perform a WP-CLI operation that only affects the DB and therefore don't require any access to "userland" files `wp-config.php` or `/wp-content/*`.

As a managed hosting provider, such operations include things like changing a domain (`wp search-replace`) or checking DB tables (`wp db check`) that only require WordPress core and a working DB connection, nothing from userland filesystem needs to be loaded. Most of which can be easily bypassed using the handy `--skip-*` globals, but the same cannot be said for the `wp-config.php` file.

It seems there should be a way to remotely manage a WordPress site with WP-CLI without the need to run a `wp config create --force` wiping away the customer's changes (regardless of whether they are good/working changes or not).

**Comment by danielbachhuber** [comment]
> we work to solve some common problems we encounter. The chief among them is customers breaking their own config files with bad syntax (or getting hacked) causing our WP-CLI calls to fail.

What sort of WP-CLI calls are you making when a config file has bad syntax or is hacked?

> The solution proposed here aims to solve for this use case (and potentially others) where admins already possess DB credentials and need to perform a WP-CLI operation that only affects the DB and therefore doesn't require any access to "userland" files `wp-config.php` or `/wp-content/*`.

If you don't need to load _any_ userland code, you could solve this problem in a much simpler way:

1. Add an extra copy of WordPress files somewhere on the server.
2. For this extra copy of WordPress, create a `wp-config.php` where the `DB_*` constants are fed from environment variables.
3. Run database manipulation commands against arbitrary WordPress installs by setting environment variables for the current session and executing WP-CLI against the extra copy of WordPress.

`wp host-check` is a command I worked on with DreamHost that includes a minimal copy of WordPress for a similar purpose https://github.com/danielbachhuber/host-check-command

I'm concerned that introducing a `WP_CONFIG_PATH` feature could easily have unintended side-effects, given:

* It's very low-level.
* It's not something that WordPress supports, so could break unexpectedly.

As such, if there's another way of solving the use case, I think we should do that.

**Comment by jonathanbardo** [comment]
Hey @danielbachhuber,

The solution you mention above is definitely working for database manipulation like Frankie mentionned. Unfortunately this needs to go above this use case and provide us with access to userland files without necessarily having to resort loading those files (for eg. `wp plugin list` and `wp plugin update --all`).

I get that it is a really low level change but I would argue that skipping plugins and themes are very similar in nature and aren't supported by WordPress either. 

Being able to simply override the wp-config.php file with a template of our choosing seems like the simplest solution for us. This template could then read from environnment variables and be able to load WordPress without any potential hiccup. This would alleviate us to have to maintain a separate installation and would require very little modification on the way we invoke wp-cli at the moment. 

So in summary, we still have to be in userland in order to read the filesystem (not from a blank WP install) we just want to prevent userland code from loading.

**Comment by danielbachhuber** [comment]
@schlessera Opinions?

**Comment by schlessera** [comment]
I had already spent some thought on this previously, and my hunch would be a different approach.

Instead of providing an alternative location to the config, I was thinking of adding support for an optional CLI-specific config file (`wp-cli-config.php`, `wp-config.cli.php` or something similar) that would take precedence if it exists. This has the following advantages:

- This does not change the actual path in which to find the config file, which would be problematic because it could be pointing to any path, inside or outside the webroot.
- Configuration is project-dependent anyway, so it makes sense to have it be part of the project structure.
- As the CLI-specific config takes precedence, it will override the existing WP config. However, you can still add an explicit include from one to the other, so you can partially override instead of completely replacing, if that is what you need.
- It allows complex projects to deal with CLI-vs-FPM architectural issues in a clean way.
- Just as the regular WP config, this could be found not only in the public root, but also one level above it for security reasons. Together with a simple `wp-cli.yml` file, you can pull this out of the web root and leave the actual WordPress file layout intact.
- As opposed to an environment variable, this can deal with multiple projects being referenced within one environment session.

@jonathanbardo, @fjarrett, I would love to know whether this would work for you. I know that this is not an exact feature match to an environment variable, but I think it could solve your pain points just as well.

**Comment by jonathanbardo** [comment]
@schlessera this is something that would be viable I think. I'm only worried that if users are aware of such a config they would probably want to play with it (and we should let them I think). What I feel like this is missing is really a config that is part of the wp-cli bootstrap process and something that the user couldn't play with (or wouldn't want to because they have another way already available)

**Comment by danielbachhuber** [comment]
Another potential option to consider: add some check in `locate_wp_config()` that bails early if `ABSPATH` is already defined. You could then load whatever code you wanted with `wp --require=` and simply define `ABSPATH` to have WP-CLI bail on trying to find a `wp-config.php` file.

**Comment by jonathanbardo** [comment]
@danielbachhuber That is a good solution as well! I prefer it actually since it doesn't introduce any other ENV var reading.

**Comment by schlessera** [comment]
@jonathanbardo Yes, I'm aware of that difference, but it might be a safe assumption that if a user adds a WP-CLI-specific configuration, it would be to make WP-CLI work as expected. If I understand correctly, your problems are more due to PHP-FPM specific entries in the config file that break on cli SAPI.

@danielbachhuber I like that idea, and I think we should consider this in addition to a separate config file. It seems this is just a sensible safeguard in any use case. I'm just wondering whether we would not hit some edge cases then where we end up not loading a config at all because ABSPATH was provided before the config loading.

**Comment by jonathanbardo** [comment]
@schlessera I think our main issue really lies that we can't really trust any kind of user config on mission critical wp-cli calls. The ABSPATH solution would solve that pretty nicely. Don't get me wrong, I think having a separate wp-cli config is great for the user and should definitely be on the roadmap.

**Comment by jonathanbardo** [comment]
@danielbachhuber only problem I'm seeing is that `ABSPATH` is already defined by `Runner::set_wp_root()` so we wouldn't be able to check if it's already defined in `locate_wp_config()` since it would already be loaded at that point anyways.

**Comment by schlessera** [comment]
Just another random thought...
What if we had `--skip-config` in addition to `--skip-plugins` and `--skip-themes`?
Skipping the config could mean that we don't try to parse the actual `wp-config.php` file, just set `ABSPATH` and then head straight into `wp-settings.php`. With a `--require` you could set the DB settings as needed.

Another thing to figure out, whatever the approach: what does this mean for the salts and is this relevant to your use cases? If you don't use the same salts, it might lead to inadvertent logouts or similar, I think.

**Comment by schlessera** [comment]
Skipping the config entirely would probably mean we will hit completely new error conditions, though...

**Comment by jonathanbardo** [comment]
This is difinitely something to only use when you know what you are doing. That's why I was thinking that an env variable might obscure this a little bit whereas `--skip-config` might be a bit too general.

**Comment by fjarrett** [comment]
Gave this some thought over the weekend and I too think that `--skip-config` would be too general here. Since a config is always required that flag would always break WordPress unless you knew to use `--require` to load something else, which feels like giving people the run-around and a PITA to document.

I think either a `--wp-config-path=` global option or `$WP_CONFIG_PATH` env variable would be the cleanest implementations to choose from. I don't have a strong preference either way, there are pros and cons to each.

I would probably lean towards the global option route simply because there is already a `--path` option which feels like a similar sibling to this. So then you'd have the ability to load a custom WP path and a custom WP config path independently, and I think seeing those options in the same call would be helpful rather than hiding the WP config path in a env variable. Food for thought.

**Comment by danielbachhuber** [comment]
I think a `WP_CONFIG_PATH` environment variable is preferable over a `--wp-config-path` global argument, given the feature is meant to be semi-secret and infrequently used.

**Comment by majick777** [comment]
This would need a corresponding value available wp-cli.yml so you could do for example:
`config-path: /wp-configs/sitename.php` 
and for site aliases
`@staging:`
  ` config-path: /wp-configs/sitename-staging.php`

I am not convinced `WP_CONFIG_PATH` environment variable should be the _only_ other way to set this, `--wp-config-path` would certainly make it more accessible, I mean why make it so difficult for end users to access even if it won't be used often... just my thought is it is a _command line_ utility after all, so having to set an environment variable separately to get the desired result seems a pain (which is actually how I ended up reading this, see [related issue 5136](https://github.com/wp-cli/wp-cli/issues/5136))

**Comment by danielbachhuber** [comment]
> `--wp-config-path` would certainly make it more accessible, I mean why make it so difficult for end users to access even if it won't be used often...

We only add global parameters sparingly because they can conflict with command parameters and receive a large amount of visibility (e.g. `wp help`). This particular enhancement wouldn't meant to be used in the normal course of everyday WP-CLI use.

**Comment by majick777** [comment]
@danielbachhuber Fair enough, I guess having the option for the YML file and the environment variable would be sufficient for both intermediate and advanced use respectively in that case.

**Comment by schlessera** [comment]
@jonathanbardo, @fjarrett Let's try to get this fixed and merged for the upcoming release v2.2.0.

In the release notes and documentation, I'll add a note that this is still to be considered experimental for the v2.2.0, as we ignore what side effects we might ultimately encounter.

Once it is in that stable release, you'll be able to run tests on your systems and gather real-world data about how it actually works, and we'll be able to discuss the long-term goal for this.

Does that sound reasonable to you?

**Comment by fjarrett** [comment]
@schlessera Sounds great, thanks!
